packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2024,60 +2140,48 @@ export class KeyringController<
    * Unlock Keyrings, decrypting the vault and deserializing all
    * keyrings contained in it, using a password or an encryption key with salt.
    *
-   * @param password - The keyring controller password.
-   * @param encryptionKey - An exported key string to unlock keyrings with.
-   * @param encryptionSalt - The salt used to encrypt the vault.
+   * @param credentials - The credentials to unlock the keyrings.
    * @returns A promise resolving to the deserialized keyrings array.
    */
   async #unlockKeyrings(
-    password: string | undefined,
-    encryptionKey?: string,
-    encryptionSalt?: string,
+    credentials:
+      | {
+          password: string;
+        }
+      | {
+          encryptionKey: string;
+          encryptionSalt?: string;
+        },
   ): Promise<{
     keyrings: { keyring: EthKeyring; metadata: KeyringMetadata }[];
     newMetadata: boolean;
   }> {
     return this.#withVaultLock(async () => {
-      const encryptedVault = this.state.vault;
-      if (!encryptedVault) {
+      if (!this.state.vault) {
         throw new Error(KeyringControllerError.VaultError);
       }
+      const parsedEncryptedVault = JSON.parse(this.state.vault);
 
-      let vault;
-      const updatedState: Partial<KeyringControllerState> = {};
-
-      if (password) {
-        const result = await this.#encryptor.decryptWithDetail(
-          password,
-          encryptedVault,
-        );
-        vault = result.vault;
-        this.#password = password;
-
-        updatedState.encryptionKey = result.exportedKeyString;
-        updatedState.encryptionSalt = result.salt;
+      if ('password' in credentials) {
+        await this.#deriveEncryptionKey(credentials.password);
       } else {
-        const parsedEncryptedVault = JSON.parse(encryptedVault);
-
-        if (encryptionSalt && encryptionSalt !== parsedEncryptedVault.salt) {
-          throw new Error(KeyringControllerError.ExpiredCredentials);
-        } else {
-          encryptionSalt = parsedEncryptedVault.salt as string;
-        }
-
-        if (typeof encryptionKey !== 'string') {
-          throw new TypeError(KeyringControllerError.WrongPasswordType);
-        }
-
-        const key = await this.#encryptor.importKey(encryptionKey);
-        vault = await this.#encryptor.decryptWithKey(key, parsedEncryptedVault);
+        this.#setEncryptionKey(
+          credentials.encryptionKey,
+          credentials.encryptionSalt || parsedEncryptedVault.salt,
+        );
+      }
 
-        // This call is required on the first call because encryptionKey
-        // is not yet inside the memStore
-        updatedState.encryptionKey = encryptionKey;
-        updatedState.encryptionSalt = encryptionSalt;
+      const encryptionKey = this.#encryptionKey?.serialized;
+      if (!encryptionKey) {
```

Comment:
Nit: This branch appears to be unreachable. Maybe it was just included to convince TypeScript that the encryption key has been set?

If that's the case, we can use an `/* istanbul ignore next */` comment here to prevent it from being counted against us in the coverage stats, and a brief comment to explain why the branch is unreachable